File: src/library_skills/installer.py
import json
import os
import shutil
from dataclasses import dataclass
from importlib import metadata, resources
from pathlib import Path

from .scanner import Skill

UNIVERSAL_SKILLS_DIR = ".agents/skills"
CLAUDE_SKILLS_DIR = ".claude/skills"
TOOL_SKILL_NAME = "library-skills"
TOOL_SKILL_MARKER = ".library-skills.json"
TOOL_SKILL_KIND = "tool-skill"


@dataclass(frozen=True)
class InstallTarget:
    """A project-level skill installation target."""

    name: str
    path: Path


@dataclass(frozen=True)
class InstalledSkill:
    """A skill directory found in an installation target."""

    name: str
    type: str
    path: Path
    target: Path | None = None
    has_skill_md: bool = False


@dataclass(frozen=True)
class ToolSkillStatus:
    """Status for the copied Library Skills tool skill in a target."""

    target: InstallTarget
    path: Path
    status: str


def get_target_dirs(
    project_root: Path, *, include_claude: bool = False
) -> list[InstallTarget]:
    """Get project-level target directories.

    The universal .agents/skills target is always included. Claude-compatible
    installs are additional and never replace the universal target.
    """
    targets = [
        InstallTarget(name="universal", path=project_root / UNIVERSAL_SKILLS_DIR)
    ]
    if include_claude:
        targets.append(
            InstallTarget(
                name="claude-compatible", path=project_root / CLAUDE_SKILLS_DIR
            )
        )
    return targets


def get_all_target_dirs(project_root: Path) -> list[InstallTarget]:
    """Get all supported project-level target directories."""
    return get_target_dirs(project_root, include_claude=True)


def get_default_install_target_dirs(project_root: Path) -> list[InstallTarget]:
    """Get interactive install target defaults from project state.

    Parent directories are enough to signal that the project uses a target. When
    neither target is present, prefer the universal .agents/skills target.
    """
    targets = get_all_target_dirs(project_root)
    by_name = {target.name: target for target in targets}
    selected: list[InstallTarget] = []

    if (project_root / ".agents").exists():
        selected.append(by_name["universal"])
    if (project_root / ".claude").exists():
        selected.append(by_name["claude-compatible"])

    if not selected:
        selected.append(by_name["universal"])
    return selected


def get_existing_target_dirs(
    project_root: Path, *, include_claude: bool = False
) -> list[InstallTarget]:
    """Get targets to inspect for existing installed skills.

    Without explicit Claude compatibility, this only returns concrete skills
    directories that already exist. With include_claude, preserve the legacy
    explicit behavior of managing both known targets.
    """
    if include_claude:
        return get_target_dirs(project_root, include_claude=True)
    return [
        target for target in get_all_target_dirs(project_root) if target.path.is_dir()
    ]


class InstallError(Exception):
    """Raised when a skill cannot be installed safely."""


class ToolSkillError(Exception):
    """Raised when the tool skill cannot be managed safely."""


def install_skill(
    skill: Skill,
    target_dir: Path,
    *,
    copy: bool = False,
) -> Path:
    """Install a single skill to the target directory.

    Returns the path to the installed skill directory.
    """
    dest = target_dir / skill.name
    source = skill.skill_dir.resolve()

    if dest.exists() or dest.is_symlink():
        if dest.is_symlink():
            dest.unlink()
        elif dest.is_file():
            raise InstallError(f"Cannot overwrite non-symlink file: {dest}")
        elif dest.is_dir():
            raise InstallError(f"Cannot overwrite non-symlink directory: {dest}")

    dest.parent.mkdir(parents=True, exist_ok=True)

    if copy:
        shutil.copytree(source, dest)
    else:
        link_target = _get_symlink_target(source=source, dest=dest)
        try:
            dest.symlink_to(link_target, target_is_directory=True)
        except OSError as e:
            raise InstallError(
                f"Could not create symlink: {e}. "
                "Use --copy if your system does not support symlinks."
            ) from e

    return dest


def _get_symlink_target(*, source: Path, dest: Path) -> Path:
    try:
        return Path(os.path.relpath(source, start=dest.parent.resolve()))
    except ValueError:
        return source


def uninstall_skill(skill_name: str, target_dir: Path) -> bool:
    """Remove an installed symlink. Returns True if something was removed."""
    dest = target_dir / skill_name
    if dest.is_symlink():
        dest.unlink()
        return True
    return False


def get_tool_skill_template() -> str:
    """Return the bundled Library Skills tool skill template."""
    return (
        resources.files("library_skills.tool_skill")
        .joinpath("SKILL.md")
        .read_text(encoding="utf-8")
    )


def get_tool_skill_version() -> str:
    """Return the installed library-skills package version for marker metadata."""
    try:
        return metadata.version("library-skills")
    except metadata.PackageNotFoundError:
        return "0.0.0"


def inspect_tool_skill(target_dir: Path) -> ToolSkillStatus:
    """Inspect the copied tool skill status for a target directory."""
    target = _target_for_path(target_dir)
    dest = target_dir / TOOL_SKILL_NAME
    marker = dest / TOOL_SKILL_MARKER
    skill_md = dest / "SKILL.md"

    if not dest.exists() and not dest.is_symlink():
        return ToolSkillStatus(target=target, path=dest, status="tool skill: missing")

    if dest.is_symlink() or not dest.is_dir():
        return ToolSkillStatus(
            target=target,
            path=dest,
            status="tool skill: blocked by hand-authored directory",
        )

    if not _is_managed_tool_skill_marker(marker):
        return ToolSkillStatus(
            target=target,
            path=dest,
            status="tool skill: invalid marker"
            if marker.exists()
            else "tool skill: blocked by hand-authored directory",
        )

    if not skill_md.is_file():
        return ToolSkillStatus(target=target, path=dest, status="tool skill: stale")

    try:
        current = skill_md.read_text(encoding="utf-8")
    except (OSError, UnicodeDecodeError):
        return ToolSkillStatus(target=target, path=dest, status="tool skill: stale")

    status = (
        "tool skill: up to date"
        if current == get_tool_skill_template()
        else "tool skill: stale"
    )
    return ToolSkillStatus(target=target, path=dest, status=status)


def install_tool_skill(target_dir: Path) -> Path:
    """Install or update the copied Library Skills tool skill in a target."""
    status = inspect_tool_skill(target_dir)
    if status.status in {
        "tool skill: blocked by hand-authored directory",
        "tool skill: invalid marker",
    }:
        raise ToolSkillError(f"Cannot overwrite {status.status}: {status.path}")

    dest = target_dir / TOOL_SKILL_NAME
    if dest.exists():
        shutil.rmtree(dest)
    dest.mkdir(parents=True, exist_ok=True)
    (dest / "SKILL.md").write_text(get_tool_skill_template(), encoding="utf-8")
    (dest / TOOL_SKILL_MARKER).write_text(
        json.dumps(
            {"kind": TOOL_SKILL_KIND, "version": get_tool_skill_version()},
            indent=2,
        )
        + "\n",
        encoding="utf-8",
    )
    return dest


def _target_for_path(target_dir: Path) -> InstallTarget:
    if target_dir.as_posix().endswith(CLAUDE_SKILLS_DIR):
        return InstallTarget(name="claude-compatible", path=target_dir)
    return InstallTarget(name="universal", path=target_dir)


def _is_managed_tool_skill_marker(marker: Path) -> bool:
    try:
        data = json.loads(marker.read_text(encoding="utf-8"))
    except (OSError, UnicodeDecodeError, json.JSONDecodeError):
        return False
    return isinstance(data, dict) and data.get("kind") == TOOL_SKILL_KIND


def list_installed_skills(target_dir: Path) -> list[InstalledSkill]:
    """List skills currently installed in the target directory."""
    results: list[InstalledSkill] = []

    if not target_dir.is_dir():
        return results

    for child in sorted(target_dir.iterdir()):
        if not child.is_dir() and not child.is_symlink():
            continue
        skill_md = child / "SKILL.md"
        if child.is_symlink():
            install_type = "symlink"
            target = child.resolve()
        else:
            install_type = "directory"
            target = None

        results.append(
            InstalledSkill(
                name=child.name,
                type=install_type,
                path=child,
                target=target,
                has_skill_md=skill_md.is_file(),
            )
        )

    return results

File: ts/src/installer.ts
import {
  cpSync,
  existsSync,
  lstatSync,
  mkdirSync,
  readdirSync,
  readFileSync,
  readlinkSync,
  rmSync,
  statSync,
  symlinkSync,
  unlinkSync,
  writeFileSync,
} from "node:fs";
import { dirname, join, relative, resolve } from "node:path";
import type { Skill } from "./scanner.js";

export const UNIVERSAL_SKILLS_DIR = ".agents/skills";
export const CLAUDE_SKILLS_DIR = ".claude/skills";
export const KIRO_SKILLS_DIR = ".kiro/skills";
export const TOOL_SKILL_NAME = "library-skills";
export const TOOL_SKILL_MARKER = ".library-skills.json";
export const TOOL_SKILL_KIND = "tool-skill";

export interface FrameworkConfig {
  name: string;
  displayName: string;
  shortName: string;
  skillsDir: string;
  detectorDir: string;
  cliFlag: string;
}

export const FRAMEWORKS: Record<string, FrameworkConfig> = {
  claude: {
    name: "claude-compatible",
    displayName: "Claude Code (.claude/skills)",
    shortName: "Claude Code",
    skillsDir: ".claude/skills",
    detectorDir: ".claude",
    cliFlag: "--claude",
  },
  kiro: {
    name: "kiro-compatible",
    displayName: "Kiro (.kiro/skills)",
    shortName: "Kiro",
    skillsDir: ".kiro/skills",
    detectorDir: ".kiro",
    cliFlag: "--kiro",
  },
};

export interface InstallTarget {
  name: string;
  path: string;
}

export interface InstalledSkill {
  name: string;
  type: "symlink" | "directory";
  path: string;
  target: string | null;
  hasSkillMd: boolean;
}

export interface ToolSkillStatus {
  target: InstallTarget;
  path: string;
  status:
    | "tool skill: missing"
    | "tool skill: up to date"
    | "tool skill: stale"
    | "tool skill: invalid marker"
    | "tool skill: blocked by hand-authored directory";
}

export class InstallError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "InstallError";
  }
}

export class ToolSkillError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ToolSkillError";
  }
}

export function getTargetDirs(
  projectRoot: string,
  options: { includeClaude?: boolean } = {},
): InstallTarget[] {
  const targets: InstallTarget[] = [
    { name: "universal", path: join(projectRoot, UNIVERSAL_SKILLS_DIR) },
  ];
  if (options.includeClaude) {
    targets.push({
      name: "claude-compatible",
      path: join(projectRoot, CLAUDE_SKILLS_DIR),
    });
  }
  return targets;
}

export function getAllTargetDirs(projectRoot: string): InstallTarget[] {
  return getTargetDirs(projectRoot, { includeClaude: true });
}

export function getDefaultInstallTargetDirs(projectRoot: string): InstallTarget[] {
  const [universal, claude] = getAllTargetDirs(projectRoot);

  const selected: InstallTarget[] = [];
  if (existsSync(join(projectRoot, ".agents"))) {
    selected.push(universal);
  }
  if (existsSync(join(projectRoot, ".claude"))) {
    selected.push(claude);
  }

  return selected.length > 0 ? selected : [universal];
}

export function getExistingTargetDirs(
  projectRoot: string,
  options: { includeClaude?: boolean } = {},
): InstallTarget[] {
  if (options.includeClaude) {
    return getTargetDirs(projectRoot, { includeClaude: true });
  }
  return getAllTargetDirs(projectRoot).filter((target) => isDirectory(target.path));
}

export function installSkill(
  skill: Skill,
  targetDir: string,
  options: { copy?: boolean } = {},
): string {
  const dest = join(targetDir, skill.name);
  const source = resolve(skill.skillDir);

  if (existsSync(dest) || isSymlink(dest)) {
    if (isSymlink(dest)) {
      unlinkSync(dest);
    } else if (lstatSync(dest).isFile()) {
      throw new InstallError(`Cannot overwrite non-symlink file: ${dest}`);
    } else if (lstatSync(dest).isDirectory()) {
      throw new InstallError(`Cannot overwrite non-symlink directory: ${dest}`);
    }
  }

  mkdirSync(dirname(dest), { recursive: true });

  if (options.copy) {
    cpSync(source, dest, { recursive: true });
  } else {
    try {
      symlinkSync(getSymlinkTarget({ source, dest }), dest, "dir");
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      throw new InstallError(
        `Could not create symlink: ${message}. Use --copy if your system does not support symlinks.`,
      );
    }
  }

  return dest;
}

export function uninstallSkill(skillName: string, targetDir: string): boolean {
  const dest = join(targetDir, skillName);
  if (isSymlink(dest)) {
    unlinkSync(dest);
    return true;
  }
  return false;
}

export function inspectToolSkill(targetDir: string, template: string): ToolSkillStatus {
  const target = targetForPath(targetDir);
  const dest = join(targetDir, TOOL_SKILL_NAME);
  const marker = join(dest, TOOL_SKILL_MARKER);
  const skillMd = join(dest, "SKILL.md");

  if (!existsSync(dest) && !isSymlink(dest)) {
    return { target, path: dest, status: "tool skill: missing" };
  }
  if (isSymlink(dest) || !isDirectory(dest)) {
    return {
      target,
      path: dest,
      status: "tool skill: blocked by hand-authored directory",
    };
  }
  if (!isManagedToolSkillMarker(marker)) {
    return {
      target,
      path: dest,
      status: existsSync(marker)
        ? "tool skill: invalid marker"
        : "tool skill: blocked by hand-authored directory",
    };
  }
  if (!existsSync(skillMd)) {
    return { target, path: dest, status: "tool skill: stale" };
  }
  const status =
    readFileSync(skillMd, "utf-8") === template
      ? "tool skill: up to date"
      : "tool skill: stale";
  return { target, path: dest, status };
}

export function installToolSkill({
  targetDir,
  template,
  version,
}: {
  targetDir: string;
  template: string;
  version: string;
}): string {
  const status = inspectToolSkill(targetDir, template);
  if (
    status.status === "tool skill: blocked by hand-authored directory" ||
    status.status === "tool skill: invalid marker"
  ) {
    throw new ToolSkillError(`Cannot overwrite ${status.status}: ${status.path}`);
  }

  const dest = join(targetDir, TOOL_SKILL_NAME);
  if (existsSync(dest)) {
    rmSync(dest, { recursive: true, force: true });
  }
  mkdirSync(dest, { recursive: true });
  writeFileSync(join(dest, "SKILL.md"), template, "utf-8");
  writeFileSync(
    join(dest, TOOL_SKILL_MARKER),
    `${JSON.stringify(
      { kind: TOOL_SKILL_KIND, version },
      null,
      2,
    )}\n`,
    "utf-8",
  );
  return dest;
}

export function listInstalledSkills(targetDir: string): InstalledSkill[] {
  if (!isDirectory(targetDir)) {
    return [];
  }

  return readdirSync(targetDir)
    .sort()
    .flatMap((name) => {
      const path = join(targetDir, name);
      if (!isSymlink(path) && !isDirectory(path)) {
        return [];
      }

      const type = isSymlink(path) ? "symlink" : "directory";
      return [
        {
          name,
          type,
          path,
          target: type === "symlink" ? resolveSymlink(path) : null,
          hasSkillMd: existsSync(join(path, "SKILL.md")),
        } satisfies InstalledSkill,
      ];
    });
}

function targetForPath(targetDir: string): InstallTarget {
  return targetDir.endsWith(CLAUDE_SKILLS_DIR)
    ? { name: "claude-compatible", path: targetDir }
    : { name: "universal", path: targetDir };
}

function isManagedToolSkillMarker(marker: string): boolean {
  try {
    const data = JSON.parse(readFileSync(marker, "utf-8")) as { kind?: unknown };
    return data.kind === TOOL_SKILL_KIND;
  } catch {
    return false;
  }
}

function getSymlinkTarget({ source, dest }: { source: string; dest: string }): string {
  const relativeTarget = relative(dirname(resolve(dest)), source);
  return relativeTarget === "" ? source : relativeTarget;
}

function resolveSymlink(path: string): string | null {
  try {
    return resolve(dirname(path), readlinkSync(path));
  } catch {
    return null;
  }
}

function isSymlink(path: string): boolean {
  try {
    return lstatSync(path).isSymbolicLink();
  } catch {
    return false;
  }
}

function isDirectory(path: string): boolean {
  try {
    return statSync(path).isDirectory();
  } catch {
    return false;
  }
}

export const testing = {
	getSymlinkTarget,
	isDirectory,
	isSymlink,
	resolveSymlink,
};
